Code/GraphMol/Depictor/RDDepictor.cpp

```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+  }
+  detail::generateDepictionMatching2DStructure(mol, reference, refMatchVect,
+                                               refMatch, confId, forceRDKit);
+}
+
 //! \brief Compute 2D coordinates where a piece of the molecule is
 //   constrained to have the same coordinates as a reference.
 void generateDepictionMatching2DStructure(RDKit::ROMol &mol,
                                           const RDKit::ROMol &reference,
                                           int confId,
                                           RDKit::ROMol *referencePattern,
-                                          bool acceptFailure, bool forceRDKit) {
+                                          bool acceptFailure, bool forceRDKit,
+                                          bool allowRGroups) {
   std::vector<int> refMatch;
   RDKit::MatchVectType matchVect;
+  RDKit::MatchVectType matchVectFiltered;
+  auto &matchVectRef = matchVect;
   if (referencePattern) {
-    if (reference.getNumAtoms(true) != referencePattern->getNumAtoms(true)) {
+    if (referencePattern->getNumAtoms(true) > reference.getNumAtoms(true)) {
       throw RDDepict::DepictException(
-          "When a pattern is provided, it must have the same number of atoms "
-          "as the reference");
+          "When a pattern is provided, its number of atoms must be "
+          "<= reference");
     }
     RDKit::MatchVectType refMatchVect;
     RDKit::SubstructMatch(reference, *referencePattern, refMatchVect);
     if (refMatchVect.empty()) {
       throw RDDepict::DepictException(
           "Reference pattern does not map to reference.");
     }
-    refMatch.reserve(refMatchVect.size());
+    refMatch.resize(reference.getNumAtoms(true), -1);
     for (auto &i : refMatchVect) {
-      refMatch.push_back(i.second);
+      refMatch[i.first] = i.second;
     }
-    RDKit::SubstructMatch(mol, *referencePattern, matchVect);
   } else {
-    refMatch.reserve(reference.getNumAtoms(true));
-    for (unsigned int i = 0; i < reference.getNumAtoms(true); ++i) {
-      refMatch.push_back(i);
-    }
-    RDKit::SubstructMatch(mol, reference, matchVect);
+    refMatch.resize(reference.getNumAtoms(true));
+    std::iota(refMatch.begin(), refMatch.end(), 0);
   }
-
-  RDGeom::INT_POINT2D_MAP coordMap;
-  if (matchVect.empty()) {
-    if (!acceptFailure) {
-      throw RDDepict::DepictException(
-          "Substructure match with reference not found.");
-    }
-  } else {
-    const RDKit::Conformer &conf = reference.getConformer(confId);
-    for (RDKit::MatchVectType::const_iterator mv = matchVect.begin();
-         mv != matchVect.end(); ++mv) {
-      RDGeom::Point3D pt3 = conf.getAtomPos(refMatch[mv->first]);
-      RDGeom::Point2D pt2(pt3.x, pt3.y);
-      coordMap[mv->second] = pt2;
+  const RDKit::ROMol &query =
+      (referencePattern ? *referencePattern : reference);
+  if (allowRGroups) {
+    std::unique_ptr<RDKit::ROMol> molHs(RDKit::MolOps::addHs(mol));
+    if (molHs) {
```

Comment:
At the moment you silently fail if this evaluates to false.
There currently isn't any circumstance under which `addHs()` can return nullptr, so I don't think the test is necessary at all, but if you want to be extra careful, maybe switch to a `CHECK_INVARIANT()`